osquery/dispatcher/bro.cpp

```diff
@@ -0,0 +1,230 @@
+ *
+ */
+
+#include <iostream>
+
+#include <unistd.h>
+
+#include <broker/broker.hh>
+#include <broker/endpoint.hh>
+#include <broker/message_queue.hh>
+
+#include <osquery/config.h>
+#include <osquery/flags.h>
+#include <osquery/logger.h>
+#include <osquery/sql.h>
+
+#include "osquery/bro/BrokerManager.h"
+#include "osquery/bro/utils.h"
+#include "osquery/dispatcher/bro.h"
+
+namespace osquery {
+
+FLAG(string, bro_ip, "localhost", "IP address of bro (default localhost)")
+
+FLAG(uint64, bro_port, 9999, "Port of bro (default 9999)")
+
+FLAG(string, bro_groups, "{}", "List of groups (default {})")
+
+FLAG(bool, disable_bro, true, "Disable bro (default true)");
+
+void BroRunner::start() {
+  // Setup Broker Endpoint
+  LOG(INFO) << "Setup Broker Manager";
+  broker::init();
+  BrokerManager* bm = BrokerManager::getInstance();
+  QueryManager* qm = QueryManager::getInstance();
+
+  // Set Broker UID
+  std::string ident;
+  auto status_huuid = getHostUUID(ident);
+  if (status_huuid.ok())
+    bm->setNodeID(ident);
+  const auto& uid = bm->getNodeID();
+
+  // Subscribe to all and individual topic
+  bm->createEndpoint(uid);
+  bm->createMessageQueue(bm->TOPIC_ALL);
+  bm->createMessageQueue(bm->TOPIC_PRE_INDIVIDUALS + uid);
+
+  // Set Broker groups and subscribe to group topics
+  std::vector<std::string> bro_groups;
+  parseBrokerGroups(FLAGS_bro_groups, bro_groups);
+  for (std::string g : bro_groups) {
+    bm->addGroup(g);
+  }
+
+  // Connect to Bro and send announce message
+  LOG(INFO) << "Connecting to '" << FLAGS_bro_ip << ":" << FLAGS_bro_port
+            << "'";
+  auto status_broker = bm->peerEndpoint(FLAGS_bro_ip, FLAGS_bro_port);
+  if (!status_broker.ok()) {
+    LOG(ERROR) << status_broker.getMessage();
+    Initializer::requestShutdown(status_broker.getCode());
+  }
+  LOG(INFO) << "Broker connection established. "
+            << "Ready to process, entering main loop.";
+
+  /*
+  *
+  * MAIN Loop
+  *
+  */
+
+  // Wait for any requests
+  fd_set fds;
+  std::vector<std::string> topics;
+  int sock;
+  broker::message_queue* queue = nullptr;
+  while (!interrupted()) {
+    // Retrieve info about each message queue
+    FD_ZERO(&fds);
+    bm->getTopics(topics); // List of subscribed topics
+    int sMax = 0;
+    for (auto topic : topics) {
+      sock = bm->getMessageQueue(topic)->fd();
+      if (sock > sMax) {
+        sMax = sock;
+      }
+      FD_SET(sock, &fds); // each topic -> message_queue -> fd
+    }
+    // Wait for incoming message
+    if (select(sMax + 1, &fds, NULL, NULL, NULL) < 0) {
+      LOG(ERROR) << "Select returned an error code";
+      continue;
+    }
+
+    // Check for the socket where a message arrived on
+    for (auto topic : topics) {
+      queue = bm->getMessageQueue(topic);
+      sock = queue->fd();
+      if (FD_ISSET(sock, &fds)) {
+        // Process each message on this socket
+        for (auto& msg : queue->want_pop()) {
```

Comment:
There's quite a bit of logic in this nested scope, from a style point of view, is it possible to refactor this out into another `static inline` method perhaps?